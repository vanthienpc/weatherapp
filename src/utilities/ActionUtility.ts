import IAction from 'models/ActionModel';
import { call, put } from 'redux-saga/effects';
import ErrorResponseModel from 'models/ErrorResponseModel';

export function* createSagaEffect<P>(
  action: (...arg: any[]) => IAction<any>,
  effect: (...arg: any[]) => Promise<P | ErrorResponseModel>,
  ...args: any[]
): Generator {
  const response = yield call(effect, ...args);
  const isError: boolean = response instanceof ErrorResponseModel;

  return yield put(action(response, isError));
}

export const createAction = <T = undefined>(
  type: string,
  payload?: T,
  error = false,
  meta: any = null,
): IAction<T> => {
  return { type, payload, error, meta };
};
