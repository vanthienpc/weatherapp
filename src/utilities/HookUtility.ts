import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
  RootStateOrAny,
  TypedUseSelectorHook,
} from 'react-redux';

export const useSelector: TypedUseSelectorHook<RootStateOrAny> = (state: RootStateOrAny) =>
  originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
