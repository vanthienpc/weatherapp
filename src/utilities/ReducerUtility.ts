import { Reducer } from 'redux';
import IAction from 'models/ActionModel';

type ReducerMethod<T> = (state: T, action: IAction<any>) => T;
type ReducerMethods<T> = { [actionType: string]: ReducerMethod<T> };

export const handleReducer = <T = any>(initialState: T, methods: ReducerMethods<T>): Reducer<T> => {
  return (state: T = initialState, action: IAction<any>): T => {
    const method: ReducerMethod<T> | undefined = methods[action.type];

    if (!method || action.error) {
      return initialState;
    }

    return method(state, action);
  };
};
