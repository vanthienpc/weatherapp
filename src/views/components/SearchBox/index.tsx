import React from 'react';
import { Input } from 'antd';
import { SearchProps } from 'antd/lib/input';
import styled from 'styled-components';
import { Dispatch } from 'redux';
import IStore from 'models/StoreModel';
import { useDispatch, useSelector } from 'utilities/HookUtility';
import { selectRequest } from 'selectors/RequestSelector';
import * as WeatherAction from 'stores/weather/WeatherAction';
import * as WeatherType from 'stores/weather/WeatherType';

const { Search } = Input;

const SearchField = styled(Search).attrs((props: SearchProps) => ({
  size: 'large',
  loading: props.loading,
}))`
  height: 46px;
`;

const SearchBox: React.FC = () => {
  const dispatch: Dispatch = useDispatch();

  const isRequest: boolean = useSelector((state: IStore) =>
    selectRequest(state, [WeatherType.FETCH_WEATHER_REQUEST]),
  );

  const onSearch = (value: string) => value && dispatch(WeatherAction.fetchWeatherRequest(value));

  return <SearchField placeholder="Enter city name..." onSearch={onSearch} loading={isRequest} />;
};

export default SearchBox;
