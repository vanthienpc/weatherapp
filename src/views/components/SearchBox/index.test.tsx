import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import createSagaMiddleware from 'redux-saga';
import configureStore from 'redux-mock-store';
import * as HookUtility from 'utilities/HookUtility';
import SearchBox from '.';

const setup = (initialState?: boolean) => {
  let store: any;

  const sagaMiddleware = createSagaMiddleware();

  store = configureStore([sagaMiddleware])(initialState);

  jest.spyOn(HookUtility, 'useSelector').mockImplementation(() => store.getState());

  jest.spyOn(HookUtility, 'useDispatch').mockImplementation(() => store.dispatch);

  //@ts-ignore
  const wrapper = shallow(<SearchBox store={store} />);

  return { wrapper, store };
};

it('render should match the snapshot', () => {
  const initialState = false;
  const { wrapper } = setup(initialState);
  expect(toJson(wrapper)).toMatchSnapshot();
});
