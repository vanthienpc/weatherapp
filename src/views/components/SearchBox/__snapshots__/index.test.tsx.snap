// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should match the snapshot 1`] = `
<Styled(Search)
  loading={false}
  onSearch={[Function]}
  placeholder="Enter city name..."
/>
`;
